### PR TITLE
CPP: Add copy ctors for std::vector and SCPStatement

### PR DESCRIPTION
**Pull request by omerfirmak**
New `dlist.d` is only there until https://github.com/dlang/phobos/pull/8315 is merged and included in a release.

### CODE REVIEW

**Comment by omerfirmak** [comment]
This compiles fine on ldc 1.27.0 :/

**Comment by omerfirmak** [comment]
This commit breaks it https://github.com/ldc-developers/phobos/commit/73e51d395ebee36029f337b3b97bd5efbcb694a4

**Comment by omerfirmak** [comment]
Now also depends on https://github.com/dlang/phobos/pull/8316

**Comment by omerfirmak** [comment]
They (https://github.com/dlang/phobos/pull/8315 and https://github.com/dlang/phobos/pull/8316) are both merged, I will rerun this when they land in `ldc-master`.

**Comment by omerfirmak** [comment]
dmd-2.098.1 is released, giving this a spin again.

**Comment by omerfirmak** [comment]
Works fine with RPC & LocalRest, but with HTTP endpoints we are not externalizing any blocks.

**Comment by omerfirmak** [comment]
Fixed the bug that caused envelopes to get partially serialized in vibe.d
But there is now a runtime crash when using HTTP interfaces between validators, RPC works fine with no crashes.

**Comment by omerfirmak** [comment]
> But there is now a runtime crash when using HTTP 

Have been working on this for the past two days, using dmd-2.098.1

[This ](https://issues.dlang.org/show_bug.cgi?id=22505)came back to bite us. Tried to fix the issue itself but turns out, it is a really involved one. Added a workaround for it in the meantime.

Marking ready for reviews before the new ldc release.

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -565,7 +565,7 @@ extern(D):
         {
             if (!proc)
             {
-                this.queued_envelopes.insertBack(cast()envelope);
+                this.queued_envelopes.insertBack(envelope.serializeFull.deserializeFull!SCPEnvelope());
```
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:571]:
```suggestion
                this.queued_envelopes.insertBack(envelope.clone());
```

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -1084,21 +1084,13 @@ extern(D):
...
-            env.statement.pledges = deserializeFull!(SCPStatement._pledges_t)(
-                serializeFull(env.statement.pledges));
-        }
-        catch (Exception ex)
-        {
-            assert(0, ex.to!string);
-        }
-
-        this.network.validators().each!(v => v.client.sendEnvelope(env));
+            copy = envelope.serializeFull.deserializeFull!SCPEnvelope();
```
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:1112]:
```suggestion
            copy = envelope.clone();
```

**Review by hewison-chris — APPROVED** [review]
LGTM, maybe make  use of clone() though :-)

**Comment by omerfirmak** [comment]
> LGTM, maybe make use of clone() though :-)

Actually we dont need serializer for deep copies anymore. We have copy ctors for all CPP types now :)

**Comment by omerfirmak** [comment]
Updated ldc to the newest release, once the alpine ldc issue is fixed we can probably finally get this merged.

File: source/scpd/Cpp.d [hunk 3]
```diff
@@ -497,8 +498,10 @@ unittest
     iota(5).each!((num) {ubyte b = cast(ubyte)num; vec_ubyte.push_back(b);});
     auto serialized = vec_ubyte.toString();
     assert(serialized == `"AAECAwQ="`, "actual serialized: " ~ serialized);
-    auto deserialized = vec_ubyte.fromString(serialized)[];
-    assert(deserialized == [0, 1, 2, 3, 4], "actual deserialized: " ~ to!string(deserialized));
+    vector!ubyte deserialized = vec_ubyte.fromString(serialized);
+    assert(deserialized[] == [0, 1, 2, 3, 4], "actual deserialized: " ~ to!string(deserialized));
```
Comment (Geod24) [inline source/scpd/Cpp.d]: You can remove the message now

File: source/scpd/types/Stellar_SCP.d [hunk 4]
```diff
@@ -102,6 +102,71 @@ struct SCPStatement {
...
+                case SCPStatementType.SCP_ST_NOMINATE:
+                    destroy(this.nominate_);
+                    break;
+            }
+        }
+
+        this (ref return scope inout(_pledges_t) rhs) inout @trusted nothrow @nogc pure
+        {
+            this.type_ = rhs.type_;
+            if (this.type_== SCPStatementType.SCP_ST_PREPARE)
```
Comment (Geod24) [inline source/scpd/types/Stellar_SCP.d:127]: `final switch` ?

File: source/scpd/types/Stellar_SCP.d [hunk 5]
```diff
@@ -102,6 +102,71 @@ struct SCPStatement {
...
+            else if (this.type_== SCPStatementType.SCP_ST_EXTERNALIZE)
+                this.externalize_ = rhs.externalize_;
+            else if (this.type_== SCPStatementType.SCP_ST_NOMINATE)
+                this.nominate_ = rhs.nominate_;
+        }
+
+        void opAssign()(auto ref _pledges_t rhs) @trusted nothrow @nogc pure
+        {
+            this.type_ = rhs.type_;
+            if (this.type_== SCPStatementType.SCP_ST_PREPARE)
```
Comment (Geod24) [inline source/scpd/types/Stellar_SCP.d:140]: Ditto `final switch`

File: source/scpd/types/Stellar_SCP.d [hunk 6]
```diff
@@ -226,76 +291,7 @@ struct SCPStatement {
...
-
-            We need to initialize using a literal to account for type
-            constructors, but we can't initialize the `union` in a generic way
-            (because we need to use a different name based on the `type`).
-            The normal solution is to put it in a `switch`, but since we
-            declare multiple variable (one per `switch` branch),
-            NRVO is disabled.
-            The solution is to use `static if` to ensure the compiler only sees
-            one temporary and does NRVO on this function, which in turn enables
-            NRVO on the caller.
```
Comment (Geod24) [inline source/scpd/types/Stellar_SCP.d:255]: I had forgotten about this black magic. I remember going through a lot of suffering to make sure it works, so I'll give it a spin tomorrow to make sure we do not regress.

File: source/scpd/Cpp.d [hunk 7]
```diff
@@ -69,7 +69,13 @@ private mixin template CPPBindingMixin (T, bool Copyable = true, bool DefaultCon
...
         }
 
-        @disable this (this);
+        // Workaround for https://issues.dlang.org/show_bug.cgi?id=22505
+        static if (Copyable)
+        this (this) @trusted nothrow @nogc pure
+        {
+            T temp = T(this);
+            copyCtorCPPObject!T(&this, &temp);
+        }
```
Comment (Geod24) [inline source/scpd/Cpp.d]: Waait what ? We have a copy ctor, a postblit should never be called.

**Review by Geod24 — COMMENTED** [review]
Need to give this some testing

File: source/scpd/Cpp.d (hunk 7 above)
Comment (omerfirmak) [reply source/scpd/Cpp.d]: `druntime` works in mysterious ways

**Comment by Geod24** [comment]
Can you make a PR with the first commit so I can do the CI juggling already ?

**Comment by omerfirmak** [comment]
Fixes #559

File: source/scpd/types/Stellar_SCP.d (hunk 4 above)
Comment (omerfirmak) [reply source/scpd/types/Stellar_SCP.d:127]: Fails to compile, a false positive tho

https://github.com/bosagora/agora/runs/4849358883?check_suite_focus=true#step:11:99

**Comment by omerfirmak** [comment]
Rebased and finally ready to be merged I think.

**Comment by omerfirmak** [comment]
@Geod24 , we need this merged to get rid of the serializer deep dups, they are costing us a lot of cpu time.

**Comment by Geod24** [comment]
Fails to build though

File: source/scpd/Cpp.d (hunk 7 above)
Comment (Geod24) [reply source/scpd/Cpp.d]: Do you know what calls it ?

Comment (omerfirmak) [reply source/scpd/Cpp.d]: Dynamic array runtime, while trying to relocate the array.

**Comment by omerfirmak** [comment]
> Fails to build though

Can't figure out why tho, we have `inout` copy ctors for `SCPStatement.pledges`, so SCPEnvelope should be copyable right?

**Comment by omerfirmak** [comment]
Dropped the last commit which was failing to compile

File: source/scpd/types/Utils.d [hunk 8]
```diff
@@ -31,14 +31,9 @@ public inout(opaque_vec!()) toVec (scope ref inout(Hash) data) nothrow @nogc
...
-    {
-        base: {
-            _start: data.ptr,
-            _end: data.ptr + data.length,
-            _end_of_storage: data.ptr + data.length,
-        },
-    };
+    inout opaque_vec!() ret;
+    foreach (elem; data)
+        ret.base.push_back(cast(ubyte) elem);
```
Comment (Geod24) [inline source/scpd/types/Utils.d:36]: This is a rather major change, as the previous version did not allocate

File: source/scpd/types/Stellar_SCP.d (hunk 4 above)
Comment (Geod24) [reply source/scpd/types/Stellar_SCP.d:127]: For ref:
```
Type 'FlashControlAPI' implements more than one interface: make sure the one describing the REST server is the first one
source/scpd/types/Stellar_SCP.d(130,17): Error: inout field `prepare_` initialization is not allowed in loops or after labels
source/scpd/types/Stellar_SCP.d(133,17): Error: inout field `confirm_` initialized multiple times
       Previous initialization is here.
source/scpd/types/Stellar_SCP.d(136,17): Error: inout field `externalize_` initialized multiple times
       Previous initialization is here.
source/scpd/types/Stellar_SCP.d(139,17): Error: inout field `nominate_` initialized multiple times
       Previous initialization is here.
ldc2 failed with exit code 1.
```

Comment (Geod24) [reply source/scpd/types/Stellar_SCP.d:127]: Yeah I think it's another instance of https://issues.dlang.org/show_bug.cgi?id=21080

File: source/scpd/types/Utils.d (hunk 8 above)
Comment (omerfirmak) [reply source/scpd/types/Utils.d:36]: Previous version was wrong tho, giving `std::vector` control of a GC allocated buffer seems dangerous.

Comment (Geod24) [reply source/scpd/types/Utils.d:36]: True, I thought there was a clever trick that was used that make this legit, but I can't find it (and everything I find confirms it was wrong). Can we at least call `reserve` before pushing bytes by bytes though ? :D

**Pull request merged**